File: 이것이 취업을 위한 코딩 테스트다/정렬/퀵_정렬/퀵_정렬.java
public class 퀵_정렬 {
  public static void quickSort(int[] arr, int start, int end) {
    if (start >= end) {
      return;
    }
    
    int pivot = start;
    int left = start + 1;
    int right = end;

    while (left <= right) {
      while (left <= end && arr[left] <= arr[pivot]) {
        left++;
      } while (right > start && arr[right] >= arr[pivot]) {
        right--;
      }

      if (left > right) {
        int temp = arr[right];
        arr[right] = arr[pivot];
        arr[pivot] = temp;
      } else {
        int temp = arr[right];
        arr[right] = arr[left];
        arr[left] = temp;
      }
    }
    quickSort(arr, start, right - 1);
    quickSort(arr, right + 1, end);
  }
  public static void main(String[] args) {
    int[] arr = {5, 7, 9, 0, 3, 1, 6, 2, 4, 8};
    quickSort(arr, 0, arr.length - 1);

    for (int num: arr) {
      System.out.print(num + " ");
    }
  }
}

File: 이것이 취업을 위한 코딩 테스트다/정렬/퀵_정렬/퀵_정렬.cpp
#include <iostream>
using namespace std;

int arr[] = {5, 7, 9, 0, 3, 1, 6, 2, 4, 8};

void quickSort(int *arr, int start, int end) {
  if (start >= end) {
    return;
  }

  int pivot = start;
  int left = start + 1;
  int right = end;

  while (left <= right) {
    while (left <= end && arr[left] <= arr[pivot]) {
      left++;
    }

    while (right > start && arr[right] >= arr[pivot]) {
      right--;
    }

    if (left > right) {
      swap(arr[pivot], arr[right]);
    } else {
      swap(arr[left], arr[right]);
    }

  }
  quickSort(arr, start, right - 1);
  quickSort(arr, right + 1, end);
}

int main() {
  int size = sizeof(arr)/sizeof(arr[0]);
  quickSort(arr, 0, size - 1);
  
  for (int i = 0; i < size; i++) {
    cout << arr[i] << ' ';
  }

  return 0;
}